File: java/solve.java
import java.util.*;
import java.util.stream.*;
import java.io.*;

public class solve {
    
    public static int[][] genGroups() {
        int[][] out = new int[27][9];
        int index;
        // Generate indicies of row
        for (int i = 0; i < 9; i++) {
            index = 0;
            for (int j = 0; j < 81; j++) {
                if (j / 9 == i) {  out[i][index] = j; index++;  }
            }
        }
        // Generate indicies of cols
        for (int i = 0; i < 9; i++) {
            index = 0;
            for (int j = 0; j < 81; j++) {
                if (j % 9 == i) {  out[i+9][index] = j; index++; } 
            }
        }
        // Generate indicies of boxs
        for (int i = 0; i < 9; i++) {
            index = 0;
            for (int j = 0; j < 81; j++) {
                if (3*(j/27) + (j%9)/3 == i) {  out[i+18][index] = j; index++; }
            }
        }
        return out;
    }
    static final int[][] GROUPS = genGroups();
    public static boolean legalBoard(List<Integer> board) {
        for (int[] g: GROUPS) {
            var counts = Arrays.stream(g).map(x -> board.get(x)).boxed()
                .collect(Collectors.groupingBy(
                            x -> x, 
                            Collectors.counting()
                        ));
            Optional<Long> max = Stream.of(1,2,3,4,5,6,7,8,9)
                .map(x -> counts.get(x))
                .filter(Objects::nonNull)
                .max(Long::compare);
            if (max.isPresent() && max.get() > 1) return false;
        }
        return true;
    }
    
    public static boolean isSolved(List<Integer> board) {
        return Collections.frequency(board, 0) == 0;
    }
    
    
    public static List<Integer> solve(List<Integer> board) {
        if (isSolved(board)) return board;
        List<Integer> boardCopy = new ArrayList<>(board);
        int currentPos = board.indexOf(0);
        for (int i = 1; i <= 9; i++) {
            board.set(currentPos, i);
            if (legalBoard(board)) {
                List<Integer> temp = solve(board);
                if (temp.size() > 0) return temp;
            }
            Collections.copy(board, boardCopy);
        }
        return new ArrayList<Integer>();
    }

    public static List<Integer> getBoard(String s) {
        return s.chars().map(x -> x - '0').boxed().collect(Collectors.toList());
    }

    public static void printBoard(List<Integer> board) {
        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) System.out.print(board.get(9*i+j));
            System.out.println();
        }
        System.out.println();
    }

    public static void main(String[] args) throws FileNotFoundException {
        Scanner sc = new Scanner(new File("../sudoku_puzzles"));
        while (sc.hasNextLine()) {
            List<Integer> board = getBoard(sc.nextLine());
            printBoard(solve(board));
        }
    }
}

File: cpp/solve.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>
#include <string>

using namespace std; 

vector<vector<int>> gen_groups() {
    vector<vector<int>> out;
    // Generate indices of rows
    for (int i = 0; i < 9; i++) {
        vector<int> temp;
        for (int j = 0; j < 81; j++) {
            if (j / 9 == i) temp.push_back(j);
        }
        out.push_back(temp);
    }
    // Generate indices of cols
    for (int i = 0; i < 9; i++) {
        vector<int> temp;
        for (int j = 0; j < 81; j++) {
            if (j % 9 == i) temp.push_back(j);
        }
        out.push_back(temp);
    }
    // Generate indices of boxs
    for (int i = 0; i < 9; i++) {
        vector<int> temp;
        for (int j = 0; j < 81; j++) {
            if (3 * (j/27) + (j%9)/3 == i) temp.push_back(j);
        }
        out.push_back(temp);
    }
    return out;
}

vector<vector<int>> GROUPS = gen_groups();

bool legal_board(vector<int> board) {
    for (vector<int> g: GROUPS) {
        vector<int> digits(10, 0);
        for (int i: g) {
            digits[board[i]]++;
        }
        if (*max_element(digits.begin() + 1, digits.end()) > 1) return false;
    }
    return true;
}

bool is_solved(vector<int> board) {
    return find(board.begin(), board.end(), 0) == board.end();
}

vector<int> get_board(string board) {
    vector<int> out;
    for (char i: board) {
        out.push_back(i - '0');
    }
    return out;
}

void print_board(vector<int> board) {
    for (int i = 0; i < 9; i++) {
        for (int j = 0; j < 9; j++) cout << board[9*i+j];
        cout << endl;
    }
    cout << endl;
}

vector<int> solve(vector<int> board) {
    if (is_solved(board)) return board;
    vector<int> board_copy(board.begin(), board.end());
    int current_pos = find(board.begin(), board.end(), 0) - board.begin();
    for (int digit = 1; digit <= 9; digit++) {
        board[current_pos] = digit;
        if (legal_board(board)) {
            auto temp = solve(board);
            if (temp.size() > 0) return temp;
        }
        board.assign(board_copy.begin(), board_copy.end());
    }
    return {};
}


int main() {
    ifstream file;
    string line;
    file.open("../sudoku_puzzles");
    while (getline(file, line)) {
        print_board(solve(get_board(line)));
    }
}
